File: src/main/java/com/fengsheng/phase/SendPhaseStart.java
package com.fengsheng.phase;

import com.fengsheng.Fsm;
import com.fengsheng.Game;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import org.apache.log4j.Logger;

/**
 * 情报传递阶段开始时，选择传递一张情报
 */
public record SendPhaseStart(Player player) implements Fsm {
    private static final Logger log = Logger.getLogger(SendPhaseStart.class);

    @Override
    public ResolveResult resolve() {
        Game game = player.getGame();
        if (player.isAlive()) {
            if (player.getCards().isEmpty()) {
                log.info(player + "没有情报可传，输掉了游戏");
                game.getDeck().discard(player.deleteAllMessageCards());
                player.setLose(true);
                player.setAlive(false);
                for (Player p : game.getPlayers())
                    p.notifyDying(player.location(), true);
                Player alivePlayer = getOnlyOneAlivePlayer(game.getPlayers());
                if (alivePlayer != null) {
                    CheckKillerWin.onlyOneAliveWinner(game, alivePlayer);
                    return new ResolveResult(null, false);
                }
                for (Player p : game.getPlayers())
                    p.notifyDie(player.location());
            }
        }
        if (!player.isAlive()) {
            return new ResolveResult(new NextTurn(player), true);
        }
        for (Player p : game.getPlayers()) {
            p.notifySendPhaseStart(20);
        }
        return null;
    }

    private static Player getOnlyOneAlivePlayer(Player[] players) {
        Player alivePlayer = null;
        for (Player p : players) {
            if (p.isAlive()) {
                if (alivePlayer == null)
                    alivePlayer = p;
                else
                    return null;
            }
        }
        return alivePlayer;
    }

    @Override
    public String toString() {
        return player + "的情报传递阶段开始时";
    }
}

File: src/main/kotlin/phase/SendPhaseStart.kt
package com.fengsheng.phase

import com.fengsheng.*
import com.fengsheng.skill.SkillId.LENG_XUE_XUN_LIAN
import org.apache.log4j.Logger

/**
 * 情报传递阶段开始时，选择传递一张情报
 */
data class SendPhaseStart(override val whoseTurn: Player) : ProcessFsm() {
    override fun onSwitch() {
        val game = whoseTurn.game!!
        game.addEvent(SendPhaseStartEvent(whoseTurn))
        for (p in game.players)
            p!!.notifySendPhaseStart()
    }

    override fun resolve0(): ResolveResult? {
        val game = whoseTurn.game!!
        if (whoseTurn.alive && whoseTurn.cards.isEmpty() && whoseTurn.findSkill(LENG_XUE_XUN_LIAN) == null) {
            log.info("${whoseTurn}没有情报可传，输掉了游戏")
            val messageCards = whoseTurn.messageCards.toTypedArray()
            whoseTurn.messageCards.clear()
            game.deck.discard(*messageCards)
            whoseTurn.lose = true
            whoseTurn.alive = false
            for (p in game.players) p!!.notifyDying(whoseTurn.location, true)
            for (p in game.players) p!!.notifyDie(whoseTurn.location)
        }
        if (!whoseTurn.alive)
            return ResolveResult(NextTurn(whoseTurn), true)
        for (p in game.players)
            p!!.notifySendPhaseStart(Config.WaitSecond)
        return null
    }

    override fun toString(): String {
        return "${whoseTurn}的情报传递阶段开始时"
    }

    companion object {
        private val log = Logger.getLogger(SendPhaseStart::class.java)
    }
}